include/seastar/core/abort_on_expiry.hh

```diff
@@ -36,6 +37,7 @@ template<typename Clock = lowres_clock>
 class abort_on_expiry {
     timer<Clock> _tr;
     seastar::abort_source _as;
+    optimized_optional<seastar::abort_source::subscription> _sub;
```

Comment:
This increases the size of the `abort_on_expiry` which is used in per-element basis by semaphore and similar. I propose to make is possible to combine abort sources in various ways externally.

Reply:
Please see alternative in https://github.com/scylladb/seastar/pull/1435

Reply:
I see this abort_on_expiry violates the principle that abort_source is an administrative thing by creating one for every semaphore entry.

Reply:
I think we abused semaphore by making it do too many things. We should have left this to user code, to use with condition_variable. That has full flexibility.

Reply:
> I see this abort_on_expiry violates the principle that abort_source is an administrative thing by creating one for every semaphore entry.

Where this principal is coming from and what it even means? What is "administrative" thing? Does aborting all requests initiated by an CQL connection on client disconnect is an "administrative" opertaion? What should be used to abort an abortable operation? Suppose we want to abort a bootstrap which consists of a lot of operations that take a semaphore. In order to make the operation abortable those waits need to be abortable as well. It means individual semaphore entries needs to be abortable.

 Why we create an abort_source per entry? Because you asked for it. Original patch allowed to make an entry either abortable or timeoutable and it accepted external abort_source as a parameter, but you wanted entries  to be simultaneously abortable and timeoutable and this is the interface we come up with.

Reply:
> I think we abused semaphore by making it do too many things. We should have left this to user code, to use with condition_variable. That has full flexibility.

It effectively means semaphore cannot be used almost anywhare. Most operations have to be abortable and timeoutable. Currently they do not since we just wait for timeout on shutdown, but with user provided timeouts it may be a long wait.